event_profiling.c

```diff
@@ -0,0 +1,415 @@
+                      json_symbol_size;
+    char *bucket_buffer = (char *)malloc(sizeof(char) * buffer_size);
+
+    /* Serialize */
+    int offset = sprintf(bucket_buffer, "[");
+    int ractors = rb_profiling_event_bucket->ractors;
+
+    offset = serialize_profiling_event_list(
+        rb_profiling_event_bucket->system_init_event_list, bucket_buffer,
+        offset);
+    for (int i = 0; i < ractors; i++)
+    {
+        rb_profiling_event_list_t *list =
+            rb_profiling_event_bucket->ractor_profiling_event_lists[i];
+        offset = serialize_profiling_event_list(list, bucket_buffer, offset);
+    }
+    int final_offset = (offset > 1) ? offset - 2 : 1;
+    sprintf(bucket_buffer + final_offset, "]\n"); /* Remove the last `,` */
+
+    /* Output to a file */
+    FILE *stream = fopen(outfile, "w");
+    fputs(bucket_buffer, stream);
+    fclose(stream);
+
+    free(bucket_buffer);
+}
+
+static inline int push_call_stack(const int index, const bool system)
+{
+    rb_profiling_event_list_t *list = select_profiling_event_list(system);
+
+    refute_greater_or_equal(list->call_stack.top,
+                            rb_event_profiling_config->max_call_stack_depth,
+                            "Too deep call stack %d\n");
+    int i = list->call_stack.top++;
+
+    list->call_stack.event_indexes[i] = index;
+    return index;
+}
+
+static inline rb_profiling_event_t *pop_call_stack(const bool system)
+{
+    rb_profiling_event_list_t *list = select_profiling_event_list(system);
+
+    int i = --(list->call_stack.top);
+    int index = list->call_stack.event_indexes[i];
+
+    rb_profiling_event_t *event = &(list->events[index]);
+    return event;
+}
+
+static bool call_stack_empty(const bool system)
+{
+    rb_profiling_event_list_t *list = select_profiling_event_list(system);
+    return list->call_stack.top == 0;
+}
+
+/* Public functions */
+
+void rb_ractor_init_profiling_event_list(rb_ractor_t *r)
+{
+    int ractor_id = r->pub.id;
+
+    int ractors = rb_profiling_event_bucket->ractors;
+    refute_greater_or_equal(ractors, rb_event_profiling_config->max_ractors,
+                            "Too many Ractors. %d \n");
+    RUBY_ATOMIC_INC(rb_profiling_event_bucket->ractors);
+
+    rb_profiling_event_list_t *list = init_profiling_event_list();
+
+    r->event_profiling_storage = list;
+
+    /* Save a pointer to serialize all events before MRI main ractor exiting */
+    rb_profiling_event_bucket->ractor_profiling_event_lists[ractor_id - 1] =
+        list;
+}
+
+int rb_trace_profiling_event_begin(const char *file, const char *func,
+                                const int line, const bool system)
+{
+    int                index = -1;
+    rb_profiling_event_t *event = get_a_profiling_event_slot(system, &index);
+
+    /* Track this event */
+    push_call_stack(index, system);
+
+    int id = get_a_new_event_id(system);
+
+    event->file = (char *)file;
+    event->function = (char *)func;
```

Comment:
```suggestion
    event->file = file;
    event->function = func;
```

Reply:
To mitigate warnings, or just exposing these warnings is okay?
```
../ruby/event_profiling.c: In function ‘rb_trace_profiling_event_begin’:
../ruby/event_profiling.c:313:17: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  313 |     event->file = file;
      |                 ^
../ruby/event_profiling.c:314:21: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  314 |     event->function = func;
```

Reply:
GCC `__FILE__` and `__func__` are constants